project/src/main/java/com/google/sps/data/PlacesFetcher.java

```diff
@@ -104,25 +107,40 @@ public PlacesFetcher(GeoApiContext geoApiContext) {
      *     for places or for places details
      */
     public ImmutableList<Place> fetch(UserPreferences preferences) throws FetcherException {
-        PlacesSearchResult[] placesSearchResult;
+        Map<String, Set<String>> placesSearchResults = new HashMap<>();
+        PlacesSearchResult[] resultsForCuisine;
+        // If user didn't choose any cuisines, search on all possible cuisines
+        Collection<String> cuisines =
+            (preferences.cuisines().isEmpty())
+                ? CUISINE_TO_SEARCH_WORDS.keySet() : preferences.cuisines();
         int attemptsCounter = 0;
         do {
-            attemptsCounter++;
-            try {
-                placesSearchResult = getPlacesSearchResults(
-                    genTextSearchRequest(preferences, INIT_SEARCH_RADIUS_M * attemptsCounter));
-            } catch (ApiException | InterruptedException | IOException | IllegalStateException e) {
-                throw new FetcherException("Couldn't fetch places from Places API", e);
+            for (String cuisine: cuisines) {
```

Comment:
This whole bit is a tad long and could be made more readable IMO. Consider the following:

1. Export the entire try / catch bit to getPlacesSearchResults and make it throw the FetcherException.
2. Use a stream, something like:

placesSearchResults = cuisines.stream().map(cuisine -> getPlacesSearchResults(..)).collect(toImmutableHashMap(cuisineResult::placeId, cuisineResult -> toImmutableHashSet(..));

Note that this is a draft and I probably missed a bit, but I recommend you check out Stream's flatMap and the collectors for all collection types (toImmutable<type>). Also, there's a groupingBy collector I think will be useful here: https://mkyong.com/java8/java-8-collectors-groupingby-and-mapping-example/. The bottom line is this whole thing (up to the while part) may be a one liner IMO.

Reply:
1. I didn't want to have any logic in the getPlacesSearchResults because this method is not tested (it's always stubed, as a solution for the query.await() method from the Places API library being untestable). So I think it is better to leave it that way regarding the try/ catch.

2. I tried different options using streams but it all seemed less readable to me.
The main problems were that:
* I need to increment the attemptsCounter for every cuisine in the stream, and the only way I found to do that is by using AtomicInteger (which I wanted to know what you think about that in general)
* I need to be able to know for each result what was the cuisine that mapped that result. So the map method will need to generate something more sophisticated then just the result, because once I use it I loose that information.
Both of these points made the .map() in the stream to have a lot of logic that I think makes it less readable then now.

We can discuss this 1:1 if you have the time :)